### DISCUSSION TITLE
OrderedDicts and tuples in custom functions

### DISCUSSION THREAD

**Discussion by lucasbracher** [zeroSteiner/rule-engine#94] [Q&A] [answered]
Good morning!

I don't know if I'm abusing the intended purpose of this project, but I created a function that recursively looks for a given key inside a JSON. The original code follows:

```
def key_exists(obj, key):
    stack = [obj]
    while stack:
        current = stack.pop()
        if isinstance(current, dict):
            if key in current:
                return True
            stack.extend(current.values())
        elif isinstance(current, list):
            stack.extend(current)
    return False
```

I tested it as a standalone function, and it worked like a charm.

However, I tested it as an embedded function in the context, and it didn't work. After some peeks and pokes, I discovered the JSON was transformed, it changed the dicts into OrderedDicts and lists into tuples. After some changes, the following code worked:

```
def key_exists(obj, key):
    stack = [obj]
    while stack:
        current = stack.pop()
        if isinstance(current, tuple):
            current = list(current)
        if isinstance(current, dict) or isinstance(current, OrderedDict):
            if key in current:
                return True
            stack.extend(list(current.values()))
        elif isinstance(current, list):
            stack.extend(current)
    return False
```

My question is, why exactly did it happen?

Thank you for the clarification!

**Comment by zeroSteiner** [comment, marked as answer]
Because I chose tuples and OrderdDict as the classes for normalizing the values that are being processed. There's a coerce value function that will perform this normalization.

**Reply by lucasbracher** [reply]
This is really interesting! Thank you!
Which were your criteria about choosing them over other possibilities, if can I ask?

**Reply by zeroSteiner** [reply]
I chose `tuple` for the ARRAY data type to ensure that it was immutable. I wanted to prevent any mistakes I could make that would result in changing the value. By using a tuple, I need make any changes explicitly.

For OrderedDict, I chose that to guarantee that it was ordered. A while back, Python made an implementation change that meant all dictionaries were ordered, but I don't think it's guaranteed to be maintained in the future.
